### PR TITLE
i1682: Install and schedule bb8 during deploy 

### PR DESCRIPTION
**Pull request by MartinEden**
https://vimc.myjetbrains.com/youtrack/issue/VIMC-1682

This also adds the ability to restore from bb8 as the initial data source. But I've only got bb8 working for the orderly volume, which is the easy bit. There are other tickets for database backup via barman, and database restore from bb8.

First merge https://github.com/vimc/montagu/pull/92 and https://github.com/vimc/bb8/pull/20, and update the montagu-bb8 submodule.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -31,3 +31,6 @@
 [submodule "submodules/shiny"]
 	path = submodules/shiny
 	url = https://github.com/vimc/montagu-shiny
+[submodule "montagu-bb8"]
```
Comment (richfitz) [inline .gitmodules]: Is there a reason, aside from `backup` being at the top level, not to put this under `submodules`?

File: src/bb8_backup.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+@requires_bb8_setup
+def backup():
+    print("Performing bb8 backup")
+    run(["bb8", "backup"], check=True)
+
+
+@requires_bb8_setup
+def schedule():
+    print("Scheduling bb8 backup")
+    run("./schedule", cwd=bb8_dir, check=True)
```
Comment (richfitz) [inline src/bb8_backup.py]: Does this mean we still need to deploy as root? What does the workflow look like if this step fails?

For barman, I allowed this part to fail and printed a message to run.  One way of managing that would be to add a flag to `schedule` that controls if it must work or not and printing the instructions there

File: src/bb8_backup.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    return decorated
+
+
+def setup():
+    global finished_setup
+    template = "- Configuring and installing bb8 backup service with these " \
+               "targets: {}"
+    print(template.format(targets))
+    args = ["./setup.sh", "../config.json"] + targets
```
Comment (richfitz) [inline src/bb8_backup.py]: this path algebra here is confusing me - this is `../config.json` from the point of view of `../montagu-bb8/bb8` which ends up being `../montagu-bb8/config.json` I think.  Would it be easier (and allow more easy moving of modules at a later date if wanted) to run the path of the config through `os.path.abspath` and pass that through as the argument?

File: src/bb8_backup.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    run(args, cwd=bb8_dir, check=True)
+    finished_setup = True
+
+
+@requires_bb8_setup
+def backup():
+    print("Performing bb8 backup")
+    run(["bb8", "backup"], check=True)
```
Comment (richfitz) [inline src/bb8_backup.py]: Out of scope for this ticket but if we provided bb8 as a python module we'd be able to centralise a lot of the logic in this file for use on subsequent projects?  If that sounds useful I can open a ticket

File: src/bb8_backup.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+
+@requires_bb8_setup
+def schedule():
+    print("Scheduling bb8 backup")
+    run("./schedule", cwd=bb8_dir, check=True)
+
+
+@requires_bb8_setup
+def restore():
```
Comment (richfitz) [inline src/bb8_backup.py]: This no longer needs `service` (vs the duplicati version) which means that the volume names are duplicated between the docker-compose.yml and the montagu-bb8 configuration https://github.com/vimc/montagu-bb8/blob/master/config.json

Can we (at some point, not necessarily here) move to generating the json used by bb8 from the configuration as used by montagu?

File: .gitmodules (hunk 1 above)
Comment (MartinEden) [reply .gitmodules]: I feel like the things in `submodules` are the things that we just include as submodules so we can do git magic and track which docker images to deploy. Whereas `backup` and `bb8` are different in that we are actually using the contents of the submodules to do things. So I wanted to keep them separate. I agree that a folder named `submodules` is not exactly the clearest way to do that, but moving submodules around is a pig.

File: src/bb8_backup.py (hunk 2 above)
Comment (MartinEden) [reply src/bb8_backup.py]: Yeah, currently, if there is anything that bb8 needs to do as root (which there will only sometimes be) I intend the deploy to fail and you to rerun it as root. I fear that a warning message will just get lost?

File: src/bb8_backup.py (hunk 3 above)
Comment (MartinEden) [reply src/bb8_backup.py]: Like this?

File: src/bb8_backup.py (hunk 4 above)
Comment (MartinEden) [reply src/bb8_backup.py]: That sounds good :)

File: src/bb8_backup.py (hunk 5 above)
Comment (MartinEden) [reply src/bb8_backup.py]: That sounds like a good idea: https://vimc.myjetbrains.com/youtrack/issue/VIMC-1693

File: .gitmodules (hunk 1 above)
Comment (richfitz) [reply .gitmodules]: yeah, that makes sense

File: src/bb8_backup.py (hunk 4 above)
Comment (richfitz) [reply src/bb8_backup.py]: https://vimc.myjetbrains.com/youtrack/issue/VIMC-1699

**Pull request merged**